Implement resource pruning in k8s plugin

Warashi opened the pull request:
**What this PR does**:

as title

**Why we need it**:

to prune the resources which removed in the git repository

**Which issue(s) this PR fixes**:

Part of #4980

**Does this PR introduce a user-facing change?**: No

- **How are users affected by this change**:
- **Is this breaking change**:
- **How to migrate (if breaking change)**:

Warashi commented on pkg/app/pipedv1/plugin/kubernetes/provider/loader.go:
```diff
@@ -85,13 +85,23 @@ func NewLoader(registry ToolRegistry) *Loader {
 	defer func() {
-		// Add builtin annotations for tracking application live state.
 		for i := range manifests {
+			// Set the namespace for all manifests.
+			manifests[i].body.SetNamespace(input.Namespace)
```
We need this to ensure the loaded manifests' namespaces are matched with applied resource.

Warashi replied:
Oops, we have to check the `input.Namespace` is empty or not.

Warashi commented on pkg/app/pipedv1/plugin/kubernetes/provider/loader.go:
```diff
@@ -85,13 +85,23 @@ func NewLoader(registry ToolRegistry) *Loader {
+			// Set the namespace for all manifests.
+			manifests[i].body.SetNamespace(input.Namespace)
+
+			// Add builtin labels and annotations for tracking application live state.
+			manifests[i].AddLabels(map[string]string{
+				LabelManagedBy:   ManagedByPiped,
+				LabelPiped:       input.PipedID,
+				LabelApplication: input.AppID,
+				LabelCommitHash:  input.CommitHash,
+			})
```
I want these labels to retrieve live resources when pruning the resources.

ffjlabo reviewed (commented):
I haven't seen everything yet, but I left some comments 🙏

Warashi commented on pkg/app/pipedv1/plugin/kubernetes/provider/kubectl.go:
```diff
@@ -230,6 +233,111 @@ func (c *Kubectl) Get(ctx context.Context, kubeconfig, namespace string, r Resou
+	if kubeconfig != "" {
+		args = append(args, "--kubeconfig", kubeconfig)
+	}
+	if namespace == "" {
+		args = append(args, "--all-namespaces")
```
The result does not include the cluster scoped resources.
I checked with the kubectl v1.32.0.

```
$ kubectl version
Client Version: v1.32.0
Kustomize Version: v5.5.0
Server Version: v1.31.0

$ kubectl get all --all-namespaces
NAMESPACE            NAME                                             READY   STATUS    RESTARTS   AGE
kube-system          pod/coredns-6f6b679f8f-czgq7                     1/1     Running   0          2d5h
kube-system          pod/coredns-6f6b679f8f-j7vcr                     1/1     Running   0          2d5h
kube-system          pod/etcd-kind-control-plane                      1/1     Running   0          2d5h
kube-system          pod/kindnet-rxcpn                                1/1     Running   0          2d5h
kube-system          pod/kube-apiserver-kind-control-plane            1/1     Running   0          2d5h
kube-system          pod/kube-controller-manager-kind-control-plane   1/1     Running   0          2d5h
kube-system          pod/kube-proxy-58jbg                             1/1     Running   0          2d5h
kube-system          pod/kube-scheduler-kind-control-plane            1/1     Running   0          2d5h
local-path-storage   pod/local-path-provisioner-57c5987fd4-52wtm      1/1     Running   0          2d5h

NAMESPACE     NAME                 TYPE        CLUSTER-IP   EXTERNAL-IP   PORT(S)                  AGE
default       service/kubernetes   ClusterIP   10.96.0.1    <none>        443/TCP                  2d5h
kube-system   service/kube-dns     ClusterIP   10.96.0.10   <none>        53/UDP,53/TCP,9153/TCP   2d5h

NAMESPACE     NAME                        DESIRED   CURRENT   READY   UP-TO-DATE   AVAILABLE   NODE SELECTOR            AGE
kube-system   daemonset.apps/kindnet      1         1         1       1            1           kubernetes.io/os=linux   2d5h
kube-system   daemonset.apps/kube-proxy   1         1         1       1            1           kubernetes.io/os=linux   2d5h

NAMESPACE            NAME                                     READY   UP-TO-DATE   AVAILABLE   AGE
kube-system          deployment.apps/coredns                  2/2     2            2           2d5h
local-path-storage   deployment.apps/local-path-provisioner   1/1     1            1           2d5h

NAMESPACE            NAME                                                DESIRED   CURRENT   READY   AGE
kube-system          replicaset.apps/coredns-6f6b679f8f                  2         2         2       2d5h
local-path-storage   replicaset.apps/local-path-provisioner-57c5987fd4   1         1         1       2d5h
```

ffjlabo commented on pkg/app/pipedv1/plugin/kubernetes/deployment/server.go:
```diff
@@ -290,17 +295,85 @@ func (a *DeploymentService) executeK8sSyncStage(ctx context.Context, lp logpersi
+		return model.StageStatus_STAGE_FAILURE
+	}
+
+	clusterLiveResources, err := kubectl.GetAllClusterScoped(ctx, deployTargetConfig.KubeConfigPath,
```
[nit] clusterLiveResources -> clusterScopedLiveResources

Warashi replied:
applied on this commit
https://github.com/pipe-cd/pipecd/pull/5455/commits/6392bca42cf4d162ca0ac380d488281d90ae1cdd

ffjlabo commented on pkg/app/pipedv1/plugin/kubernetes/provider/resource.go:
```diff
@@ -29,42 +31,101 @@ const (
+		}
+
+		for _, r := range namespacedLiveResources {
+			ns := r.Key().namespace
+			k := r.Key().normalize()
+			if _, ok := keys[k]; !ok {
+				// The namespace should be set to the live resource's value.
+				k.namespace = ns
+				removeKeys = append(removeKeys, k)
+			}
```
Can we just use the original key?

```suggestion
			if _, ok := keys[r.Key().normalize()]; !ok {
				removeKeys = append(removeKeys, r.Key())
			}
```

Warashi replied:
applied on this commit
https://github.com/pipe-cd/pipecd/pull/5455/commits/69ac113977ead6b92164482653b2709e70408ac7

ffjlabo commented on pkg/app/pipedv1/plugin/kubernetes/provider/resource.go:
```diff
@@ -29,42 +31,101 @@ const (
+	)
+
+	{
+		keys := make(map[ResourceKey]struct{}, len(manifests))
```
[IMO] How about changing `keys` -> `normalizedKeys` to distinguish the original key?

Warashi replied:
applied on this commit
https://github.com/pipe-cd/pipecd/pull/5455/commits/e1dec976382307bd10e96e3dcfc262cbdd9e19b9

ffjlabo commented on pkg/app/pipedv1/plugin/kubernetes/provider/resource.go:
```diff
@@ -29,42 +31,101 @@ const (
+	}
+
+	{
+		keys := make(map[ResourceKey]struct{}, len(manifests))
```
> [IMO] How about changing keys -> normalizedKeys to distinguish the original key?

Same as above

Warashi replied:
applied on this commit
https://github.com/pipe-cd/pipecd/pull/5455/commits/e1dec976382307bd10e96e3dcfc262cbdd9e19b9

Warashi commented on pkg/app/pipedv1/plugin/kubernetes/provider/kubectl.go:
```diff
@@ -230,6 +233,111 @@ func (c *Kubectl) Get(ctx context.Context, kubeconfig, namespace string, r Resou
+// It runs the `kubectl api-resources` command with the specified kubeconfig and returns the
+// names of the resources that support the "list", "get", and "delete" verbs.
+func (c *Kubectl) getAPIResources(ctx context.Context, kubeconfig string) ([]string, error) {
+	args := []string{"api-resources", "--namespaced=false", "--verbs=list,get,delete", "--output=name"}
```
I fixed this on this commit.
https://github.com/pipe-cd/pipecd/pull/5455/commits/c2a747f7e47f5753e9fa3497a0b1fa313d068d47

Warashi commented on pkg/app/pipedv1/plugin/kubernetes/provider/resource.go:
```diff
@@ -29,42 +31,101 @@ const (
-	name       string
+	groupKind schema.GroupKind
+	namespace string
+	name      string
```
I added comments on this commit.
https://github.com/pipe-cd/pipecd/pull/5455/commits/6fb0aab7b4e7d325ef1e6f3fa6c0c31a854a5ddc

ffjlabo commented on pkg/app/pipedv1/plugin/kubernetes/provider/resource.go:
```diff
@@ -29,42 +31,101 @@ const (
+	{
+		keys := make(map[ResourceKey]struct{}, len(manifests))
+		for _, m := range manifests {
+			keys[m.Key().normalize().withoutNamespace()] = struct{}{}
```
[nit] It would be nice to add comment to show why using `withoutNamespace()`

Warashi replied:
I added comments on this commit.
https://github.com/pipe-cd/pipecd/pull/5455/commits/e1dec976382307bd10e96e3dcfc262cbdd9e19b9

ffjlabo commented on pkg/app/pipedv1/plugin/kubernetes/provider/resource.go:
```diff
@@ -29,42 +31,101 @@ const (
 	return k
 }
+
+func normalizeGroupKind(gk schema.GroupKind) schema.GroupKind {
+	gk.Group = strings.ToLower(gk.Group)
+	gk.Kind = strings.ToLower(gk.Kind)
+	return gk
+}
```
[imo] How about unifying to `normalize` to avoid using only this method in other code?

Warashi replied:
applied on this commit.
https://github.com/pipe-cd/pipecd/pull/5455/commits/bcfd121f95c074895e5a1986232b54501574e3fc

ffjlabo reviewed (commented):
Sorry for the more comments 🙏

ffjlabo commented on pkg/app/pipedv1/plugin/kubernetes/deployment/server_test.go:
```diff
@@ -246,12 +256,380 @@ func TestDeploymentService_executeK8sSyncStage_withInputNamespace(t *testing.T)
+
+	// The service should be created in the new namespace
+	service, err = dynamicClient.Resource(schema.GroupVersionResource{Group: "", Version: "v1", Resource: "services"}).Namespace("test-2").Get(context.Background(), "simple", metav1.GetOptions{})
+	require.NoError(t, err)
```
[imo] If possible, it would be nice to check the error is not found. (I'm not sure about the spec for the dynamic client 🙏

Warashi replied:
Thank you. I added the assertion on this commit.
https://github.com/pipe-cd/pipecd/pull/5455/commits/b12e31770d4fb4e1201c0dda0cef3f4ffa62d921

ffjlabo commented on pkg/app/pipedv1/plugin/kubernetes/deployment/server_test.go:
```diff
@@ -246,12 +256,380 @@ func TestDeploymentService_executeK8sSyncStage_withInputNamespace(t *testing.T)
+	service, err = dynamicClient.Resource(schema.GroupVersionResource{Group: "", Version: "v1", Resource: "services"}).Namespace("test-2").Get(context.Background(), "simple", metav1.GetOptions{})
+	require.NoError(t, err)
+
+	require.Equal(t, "piped", service.GetLabels()["pipecd.dev/managed-by"])
+	require.Equal(t, "piped-id", service.GetLabels()["pipecd.dev/piped"])
+	require.Equal(t, "app-id", service.GetLabels()["pipecd.dev/application"])
+	require.Equal(t, "0012345678", service.GetLabels()["pipecd.dev/commit-hash"])
```
[imo] In addition to them, we need to check the resource key to check whether the namespace has changed correcty.

Warashi replied:
Thank you. I added resource key assertion on this commit.
https://github.com/pipe-cd/pipecd/pull/5455/commits/d75fd949bb568401b9b176a3c24217ecde3c3254

ffjlabo commented:
Also imo about the test 🙏 
- How about using `t.Run` for each preparating or checking.
- It would be nice to add the testcase for the custom resouce's ones
  - When the Custom Resource's preferred version is `v1` and the actual resource's version is `v1beta1`, Is it possible to remove the resource?

That's all for now 🙏

Warashi commented on pkg/app/pipedv1/plugin/kubernetes/deployment/server_test.go:
```diff
@@ -246,12 +256,380 @@ func TestDeploymentService_executeK8sSyncStage_withInputNamespace(t *testing.T)
+
+	running := filepath.Join("./", "testdata", "prune", "running")
+
+	// read the running application config from the example file
```
Thank you. applied on this commit.
https://github.com/pipe-cd/pipecd/pull/5455/commits/f47da4b51cefa09a63fa32a8f47ec1a74e5eee6e

Warashi commented on pkg/app/pipedv1/plugin/kubernetes/deployment/server_test.go:
```diff
@@ -246,12 +256,380 @@ func TestDeploymentService_executeK8sSyncStage_withInputNamespace(t *testing.T)
+
+	target := filepath.Join("./", "testdata", "prune", "target")
+
+	// read the running application config from the example file
```
Thank you. applied on this commit.
https://github.com/pipe-cd/pipecd/pull/5455/commits/f47da4b51cefa09a63fa32a8f47ec1a74e5eee6e

Warashi commented:
@ffjlabo Thanks.

> * How about using `t.Run` for each preparating or checking.

I fixed the tests on this commit.
https://github.com/pipe-cd/pipecd/pull/5455/commits/34c7807b2771309ebad439d63445e6687cf92875

> * It would be nice to add the testcase for the custom resouce's ones
>   * When the Custom Resource's preferred version is `v1` and the actual resource's version is `v1beta1`, Is it possible to remove the resource?

I added v1beta1 pruning assertions in the cluster-scoped test because it already uses the custom resources.
https://github.com/pipe-cd/pipecd/pull/5455/commits/40af37c9b1f274beba4b783e6af2592716bdea64

Warashi commented on pkg/app/pipedv1/plugin/kubernetes/deployment/server_test.go:
```diff
@@ -136,6 +136,8 @@ func setupTestPluginConfigAndDynamicClient(t *testing.T) (*config.PipedPlugin, d
 }
 
 func TestDeploymentService_executeK8sSyncStage(t *testing.T) {
+	t.Parallel()
```
I think the operation collision does not occur because the envtest's dummy cluster is launched in each test.

ffjlabo replied:
I see. It is in `setupTestPluginConfigAndDynamicClient`. Thanks, I got your point.

ffjlabo commented:
@Warashi Thank you for the great contribution! Those fixes look good!

BTW, I noticed that it might also be nice to add a test case when updating the resource's API version and deleting it (e.g., HPA autoscaling/v1 -> autoscaling/v2, and later remove the resource).
But it might take some more time to do. Is it worth doing now? WDYT?

Warashi commented:
@ffjlabo Thank you!

> BTW, I noticed that it might also be nice to add a test case when updating the resource's API version and deleting it (e.g., HPA autoscaling/v1 -> autoscaling/v2, and later remove the resource). But it might take some more time to do. Is it worth doing now? WDYT?

This PR is already large, so I want to do it on another PR.

ffjlabo commented:
@Warashi OK! Thanks!

ffjlabo reviewed (approved):
LGreatTM 👍
